### PR TITLE
Kal input

### PR DESCRIPTION
**Pull request by ydomenjoud**
add kal-input in library

### CODE REVIEW

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# kal-input
+
+
+* should display current value
+* should add a placeholder if provided
+* should manage type (number, phone, string, currency)
+* Doit limiter le nombre de caractère en fonction de la valeur donnée en paramètre
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md]: english please

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/format/currency.format.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      }
+
+      const formattedNumber = parseFloat(clearedValue)
+        .toFixed(2) // format to number
+        .replace(/(\d)(?=(\d{3})+(?!\d))/g, '$1 '); // use space as a separator
+
+      // replace decimal point by a comma
+      return ('' + formattedNumber).replace('.', ',');
+
+    } else {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/format/currency.format.ts]: You don't need the `else`. Just `return '';`

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/format/number.format.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+export class NumberFormat extends InputFormater {
+
+  /**
+   * @inheritDoc
+   */
+  toCode(value: string): any {
+    value = (value + '')
+      .replace(/[^0-9\.\,]/g, '') // keep only numbers
+      .replace(',', '.'); // replace comma by decimal point
+    // return +value || 0;
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/format/number.format.ts]: remove code

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.spec.ts [hunk 4]
```diff
@@ -1,25 +1,172 @@
...
+  }));
+
+
+  it('should add an icon to clear field', () => {
+    const text = 'abcdefgh';
+    component.inputComponent.clearable = true;
+    fixture.detectChanges();
+    component.value = text;
+    expect(component.inputComponent.value).toEqual(text);
+    const icon = fixture.debugElement.query(By.css('kal-icon'));
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.spec.ts]: By.directive(KalIconComponent)

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts [hunk 5]
```diff
@@ -1,5 +1,14 @@
...
+import { AbstractControl, FormControl, ValidationErrors, ValidatorFn, Validators } from '@angular/forms';
+import { coerceBooleanProperty } from '@angular/cdk/coercion';
+
 import { buildProviders, FormElementComponent } from '../../utils/index';
+import { InputFormater } from './format/input-formater';
+import { NumberFormat } from './format/number.format';
+import { CurrencyFormat } from './format/currency.format';
+import { PhoneFormat } from './format/phone.format';
+import { StringFormat } from './format/string.format';
+import { of } from 'rxjs';
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts]: import

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts [hunk 6]
```diff
@@ -11,11 +20,107 @@ import { buildProviders, FormElementComponent } from '../../utils/index';
...
+
+  @Input() limit: number;
+
+  control: FormControl;
+
+  private updateOnEvent: 'change' | 'blur' | 'submit' = 'change';
+
+  private isClearable = false;
+
+  constructor(public cdr: ChangeDetectorRef) {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts]: private ?

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts [hunk 7]
```diff
@@ -11,11 +20,107 @@ import { buildProviders, FormElementComponent } from '../../utils/index';
...
+    'text': new StringFormat(),
+    'password': new StringFormat()
+  };
+
+  /**
+   * type of input  ( text, password, email, number, ... )
+   */
+  @Input() type = 'text';
+
+  @Input() limit: number;
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts:42]: You should describe all functions and class members in this component

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md (hunk 1 above)
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md]: Readme in english

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+* Doit gérer une limite lorsque le champ est de type text
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/README.md]: English again

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts [hunk 9]
```diff
@@ -11,11 +20,107 @@ import { buildProviders, FormElementComponent } from '../../utils/index';
...
+  private updateOnEvent: 'change' | 'blur' | 'submit' = 'change';
+
+  private isClearable = false;
+
+  constructor(public cdr: ChangeDetectorRef) {
     super();
   }
 
+  clearField() {
+    this.control.setValue('');
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts]: you can use this.control.reset() in this method

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts [hunk 10]
```diff
@@ -11,11 +20,107 @@ import { buildProviders, FormElementComponent } from '../../utils/index';
...
+   * get formater for this type
+   */
+  get formater(): InputFormater {
+    return KalInputComponent.formatersList[this.type] || KalInputComponent.formatersList['text'];
+  }
+
+  /**
+   * @inheritDoc
+   */
+  writeValue(value) {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.ts:97]: super.writeValue(value)

File: projects/kalidea/kaligraphi/src/lib/utils/helpers/phone.spec.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
+import { formatPhoneNumber } from './phone';
+
+describe('Helper phone', () => {
+
+  it('should format standard number', () => {
+    expect(formatPhoneNumber('0383838383')).toEqual('03 83 83 83 83');
+  });
+
+  it('should format international number with +', () => {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/utils/helpers/phone.spec.ts:9]: +33 3 83 83 83 83

File: projects/kalidea/kaligraphi/src/lib/utils/helpers/phone.spec.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    expect(formatPhoneNumber('+33383838383')).toEqual('+333 83 83 83 83');
+  });
+
+  it('should format international number with 00', () => {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/utils/helpers/phone.spec.ts:13]: 0033 3 83 83 83 83

File: src/app/atoms/input/input.component.ts [hunk 13]
```diff
@@ -9,6 +9,10 @@ import { Component, OnInit, ViewEncapsulation, ChangeDetectionStrategy } from '@
 })
 export class InputComponent implements OnInit {
 
+  type = 'text';
```
Comment (MichaelGrosio) [inline src/app/atoms/input/input.component.ts:12]: Add comment

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.spec.ts [hunk 14]
```diff
@@ -1,25 +1,172 @@
...
+  }));
+
+  it('format currency on patch value with wrong value', ((done) => {
+    component.type = 'currency';
+    fixture.detectChanges();
+
+    const userInput = '12a';
+    component.valueChanges.pipe(take(1)).subscribe(value => {
+      expect(value).toBe(0, 'emited value should be formatted in number');
+      // component.inputComponent.focusOut();
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-input/kal-input.component.spec.ts]: Comment

**Review by jsp-up — DISMISSED** [review]

**Review by MichaelGrosio — APPROVED** [review]

**Pull request merged**